File: LiteDb.Studio.Avalonia/ViewModels/MainWindowViewModel.cs
using System.Collections.ObjectModel;
using System.Reactive;
using LiteDB;
using LiteDb.Studio.Avalonia.Core;
using LiteDb.Studio.Avalonia.Infra;
using ReactiveUI;

namespace LiteDb.Studio.Avalonia.ViewModels;

public class MainWindowViewModel : ViewModelBase
{
    private ScriptViewModel _selectedTab;
    private readonly ObservableCollection<ScriptViewModel> _scriptTabs = new();
    public ObservableCollection<DbItem> DbItems { get; } = new();

    public MainWindowViewModel()
    {
        OpenNewTabCommand = ReactiveCommand.Create(OpenNewTabCommandRun);
    }

    private void OpenNewTabCommandRun()
    {
        var root = DbItems.FirstOrDefault();
        if (root is not DbFileItem f) return;
        OpenNewTab(() => f.LiteDb, f.ConnectionString.Filename, "<Todo>");
    }

    private void OpenNewTab(System.Func<LiteDatabase> getLiteDb, string dbFile, string tableName)
    {
        var scriptName = Utils.GetScriptName(_scriptTabs.Select(c => c.Header));
        var tab = new ScriptViewModel(getLiteDb, dbFile, scriptName)
        {
            Query = Utils.GetDefaultSql(tableName)
        };
        _scriptTabs.Add(tab);
        SelectedTab = tab;
    }

    private DbItem CreateTableItem(System.Func<LiteDatabase> getLiteDb, string dbFile, string tableName)
    {
        var temp = new DbItem { Title = tableName, IsCollection = true };
        var newTabAction = new DbAction(() => OpenNewTab(getLiteDb, dbFile, tableName))
        {
            Header = "open new tab"
        };
        temp.ContextMenu.Add(newTabAction);
        return temp;
    }

    public ReactiveCommand<Unit, Unit> OpenNewTabCommand { get; }

    public ObservableCollection<ScriptViewModel> Tabs => _scriptTabs;

    public ScriptViewModel SelectedTab
    {
        get => _selectedTab;
        set => this.RaiseAndSetIfChanged(ref _selectedTab, value);
    }

    public void Connect(ConnParamType con)
    {
        var dbFile = con.DbFile;
        var conString = ConnectionParameters.BuildConString(con);
        var liteDb = DbUtils.GetDb(conString);
        var name = Path.GetFileName(dbFile);
        var root = new DbFileItem(liteDb, conString)
        {
            Title = name,
            IsExpanded = true
        };

        var system = new DbItem { Title = "System" };
        root.Children.Add(system);

        var getLiteDb = () => root.LiteDb;

        var systemTables = DbUtils.GetSystemTables(liteDb);
        foreach (var doc in systemTables)
        {
            var item = new DbItem
            {
                Title = doc["name"].AsString,
                IsCollection = true
            };
            system.Children.Add(item);
        }

        var collections = DbUtils.GetCollectionNames(liteDb);
        foreach (var nameItem in collections)
        {
            var item = CreateTableItem(getLiteDb, dbFile, nameItem);
            root.Children.Add(item);
        }

        if (collections.Any())
            OpenNewTab(getLiteDb, dbFile, collections.First());
        else if (system.Children.Any())
            OpenNewTab(getLiteDb, dbFile, system.Children.First().Title);

        DbItems.Add(root);
        SelectedTab = _scriptTabs.FirstOrDefault();
    }
}

File: LiteDb.OneBella/ViewModels/MainWindowViewModel.fs
namespace OneBella.ViewModels

open System.Collections.ObjectModel
open System.IO
open LiteDB
open LiteDb.Studio.Avalonia.Core
open LiteDb.Studio.Avalonia.Infra
open ReactiveUI

type MainWindowViewModel() as this =
    inherit ViewModelBase()

    let scriptTabs = ObservableCollection<ScriptViewModel>()
    let mutable selectedTab = Unchecked.defaultof<ScriptViewModel>

    let openNewTab getLiteDb dbFile tableName =
        let scriptName = scriptTabs |> Seq.map (fun c -> c.Header) |> Utils.GetScriptName
        let tab = new ScriptViewModel(getLiteDb, dbFile, scriptName)
        tab.Query <- Utils.GetDefaultSql tableName
        scriptTabs.Add(tab)
        this.SelectedTab <- tab


    let createTableItem getLiteDb dbFile tableName =
        let temp = DbItem(Title = tableName, IsCollection = true)

        let newTabAction =
            DbAction((fun () -> openNewTab getLiteDb dbFile tableName), Header = "open new tab")

        temp.ContextMenu.Add(newTabAction)

        temp

    let openNewTabCommand =
        let run () =
            let root: DbItem = this.DbItems |> Seq.head
            let f = root :?> DbFileItem
            openNewTab (fun () -> f.LiteDb) f.ConnectionString.Filename "<Todo>"

        ReactiveCommand.Create(run)

    member x.OpenNewTabCommand = openNewTabCommand
    member val DbItems = ObservableCollection<DbItem>()
    member x.Tabs = scriptTabs

    member x.SelectedTab
        with get () = selectedTab
        and set v = x.RaiseAndSetIfChanged(&selectedTab, v) |> ignore


    member x.Connect(con: ConnParamType) =
        let dbFile = con.DbFile
        let conString = ConnectionParameters.BuildConString con

        let liteDb = DbUtils.GetDb conString

        let name = Path.GetFileName dbFile
        let root = DbFileItem(liteDb, conString, Title = name, IsExpanded = true)

        let system = DbItem(Title = "System")
        root.Children.Add system
        let getLiteDb = fun () -> root.LiteDb

        liteDb
        |> DbUtils.GetSystemTables
        |> Seq.map (fun doc -> DbItem(Title = doc.["name"].AsString, IsCollection = true))
        |> Seq.iter (fun i -> system.Children.Add i)

        let collections = DbUtils.GetCollectionNames liteDb

        collections
        |> Seq.map (fun name -> createTableItem getLiteDb dbFile name)
        |> Seq.iter (fun item -> root.Children.Add item)

        if (Seq.length collections) > 0 then
            openNewTab getLiteDb dbFile (Seq.head collections)
        else
            openNewTab getLiteDb dbFile (Seq.head system.Children).Title

        x.DbItems.Add root
        x.SelectedTab <- scriptTabs[0]
